dm/pkg/checker/privilege.go

```diff
@@ -122,72 +135,97 @@ func (pc *SourceReplicatePrivilegeChecker) Name() string {
 
-// TODO: if we add more privilege in future, we might add special checks (globally granted?) for that new privilege.
-func verifyPrivileges(result *Result, grants []string, expectedGrants map[mysql.PrivilegeType]struct{}) {
-	result.State = StateFailure
+func verifyPrivileges(result *Result, grants []string, lackGrants map[mysql.PrivilegeType]map[string]map[string]struct{}) *Error {
 	if len(grants) == 0 {
-		result.Errors = append(result.Errors, NewError("there is no such grant defined for current user on host '%%'"))
-		return
-	}
-
-	var (
-		user       string
-		lackGrants = make(map[mysql.PrivilegeType]struct{}, len(expectedGrants))
-	)
-	for k := range expectedGrants {
-		lackGrants[k] = struct{}{}
+		return NewError("there is no such grant defined for current user on host '%%'")
 	}
 
+	p := parser.New()
 	for i, grant := range grants {
-		// get username and hostname
-		node, err := parser.New().ParseOneStmt(grant, "", "")
+		if len(lackGrants) == 0 {
+			break
+		}
+		node, err := p.ParseOneStmt(grant, "", "")
 		if err != nil {
-			result.Errors = append(result.Errors, NewError(errors.Annotatef(err, "grant %s, grant after replace %s", grants[i], grant).Error()))
-			return
+			return NewError(errors.Annotatef(err, "grant %s, grant after replace %s", grants[i], grant).Error())
 		}
 		grantStmt, ok := node.(*ast.GrantStmt)
 		if !ok {
 			switch node.(type) {
 			case *ast.GrantProxyStmt, *ast.GrantRoleStmt:
 				continue
 			default:
-				result.Errors = append(result.Errors, NewError("%s is not grant statement", grants[i]))
-				return
+				return NewError("%s is not grant statement", grants[i])
 			}
 		}
 
 		if len(grantStmt.Users) == 0 {
-			result.Errors = append(result.Errors, NewError("grant has no user %s", grantStmt.Text()))
-			return
-		} else if user == "" {
-			// show grants will only output grants for requested user
-			user = grantStmt.Users[0].User.Username
+			return NewError("grant has no user %s", grantStmt.Text())
 		}
 
-		for _, privElem := range grantStmt.Privs {
-			if privElem.Priv == mysql.AllPriv {
-				if grantStmt.Level.Level == ast.GrantLevelGlobal {
-					result.State = StateSuccess
-					return
-				}
-				// REPLICATION CLIENT, REPLICATION SLAVE, RELOAD should be global privileges,
-				// thus a non-global GRANT ALL is not enough
-				for expectedGrant := range lackGrants {
-					if _, ok := privNeedGlobal[expectedGrant]; !ok {
-						delete(lackGrants, expectedGrant)
+		dbName := grantStmt.Level.DBName
+		tableName := grantStmt.Level.TableName
+		switch grantStmt.Level.Level {
+		case ast.GrantLevelGlobal:
+			for _, privElem := range grantStmt.Privs {
+				switch privElem.Priv {
+				case mysql.AllPriv:
+					return nil
+				default:
+					// mysql> show master status;
+					// ERROR 1227 (42000): Access denied; you need (at least one of) the SUPER, REPLICATION CLIENT privilege(s) for this operation
+					if privElem.Priv == mysql.SuperPriv {
+						delete(lackGrants, mysql.ReplicationClientPriv)
```

Comment:
and `replication slave`?

Reply:
`super` can not be instead of `replication slave`, I remember.

Reply:
```sql
mysql> show slave status;
ERROR 1227 (42000): Access denied; you need (at least one of) the SUPER, REPLICATION CLIENT privilege(s) for this operation
mysql> show slave hosts;
ERROR 1227 (42000): Access denied; you need (at least one of) the REPLICATION SLAVE privilege(s) for this operation
mysql> 
```

We used `replication slave` to `show slave hosts`, this command needn't `super` privilege.